Add /getglobal command to request global staff help

panpawn opened the pull request:
This is a WIP - as per the policy discussion thread in the staff forums.

This allows for trusted users to request for help with an issue that's going on in a specific room, with a brief explanation as to what's going on.

There is also a cooldown period - someone can only use this command once every two minutes at the most to prevent it from being spammed.

panpawn commented:
Still working on the UI :p

sparkychildcharlie commented on chat-commands.js:
```diff
@@ -1409,6 +1410,37 @@ exports.commands = {
+		if (target.startsWith(separator)) {
+			if (!this.can('lock')) return;
+			const requestNumber = target.substr(separator.length, target.length);
+			if (!staff.requests || !staff.requests[requestNumber]) return this.errorReply("Werid, this request doesn't seem to exist.");
```
"Werid"

panpawn replied:
Weird

sparkychildcharlie commented on chat-commands.js:
```diff
@@ -1409,6 +1410,37 @@ exports.commands = {
+			staff.add(Chat.html`|uhtmlchange|staff-request-${requestNumber}|<div class="infobox">[Help Request] <button class="button disabled" style="font-weight: bold; color: #575757; font-weight: bold; background-color: #d3d3d3">${request.roomId} (${user.name} answered)</button> - ${request.by} said: ${request.message}</div>`).update();
+		} else {
+			if (!user.trusted) return this.errorReply("Only trusted users are allowed to use this command. If you need help or a global staff member, ask the Help room.");
+			if (user.lastStaffRequest &&  Date.now() - user.lastStaffRequest < STAFF_REQUEST_THROTTLE) return this.errorReply("You cannot submit another global staff request so soon.");
```
Shouldn't this be a per-room basis - better if 4 trusted users don't use it at the same time unknowingly?

panpawn replied:
Good idea, this is done

Zarel commented:
If it's restricted to trusted users, I'm not sure the throttle is necessary...

AustinXII commented:
Staff defiently thinks the throttle is necessary, also this is still being discussed if you weren't aware @Zarel

bumbadadabum commented:
I don't see why there should be a throttle. If people abuse this they should not be public roomstaff.

sparkychildcharlie commented:
@bumbadadabum a per room throttle would be useful in preventing multiple roomauth from unknowingly sending multiple requests for staff

bumbadadabum commented:
I think the solution to that is to modnote this command's usage in the room itself, @sparkychildcharlie

panpawn commented:
Not all trusted users can see modnotes, though

sparkychildcharlie commented on chat-commands.js:
```diff
@@ -1409,6 +1410,37 @@ exports.commands = {
+				roomId: room.id,
+				message: target,
+			};
+			staff.add(Chat.html`|uhtml|staff-request-${id}|<div class="infobox">[Help Request] <button class="button" name="send" value="/getstaff ${separator}${id}">${room.id}</button> - ${user.name} said: ${target}</div>`).update();
```
Maybe use the new and nifty ``|notify|`` message type to highlight all staff in the staff room?

panpawn replied:
No, this 100% shouldn't highlight anyone

AustinXII replied:
^^^^^^^^^^^

panpawn commented:
Updated with:

- Changed the throttle to be per-room, and shortened it to a minute. This will help prevent 2 people from reporting the same problem in the same room in a short period of time.
- Changed the UI for the button to look like regular text
- Changed the command name to ``/requestglobal`` and ``/getglobal``, and updated help command accordingly

Morfent commented:
I'm not sure how I feel about `requests` being tacked on to the staff room object like this, since it'd further add to the polymorphism functions taking rooms as arguments have since they already have a number of different shapes. Depending on how much this command gets used, it may not be that big a deal, but if it gets used a lot then I think it'd be more optimal to make this a chat plugin or work it into some other part of PS' codebase

Morfent commented on chat-commands.js:
```diff
@@ -1409,6 +1410,52 @@ exports.commands = {
+		target = target.trim();
+		const separator = 'answer request id ';
+		const astext = 'text-decoration: underline; color: #0000EE; border: none; background: none; padding: 0; font-family: Verdana,sans-serif; font-size: 12px';
+		if (target.startsWith(separator)) {
```
I'm not sure using a separator like this is the most intuitive way to go about this. Maybe something like this would be better:
```javascript
if (!isNaN(+target) && +target >= 0) {
    if (!this.can('lock')) return this.errorReply("Some appropriate error message");
    ...
}
```

panpawn commented:
>I'm not sure how I feel about requests being tacked on to the staff room object like this

To be fair, I do handle garbage collection of deleting requests that have been taken care of

Morfent commented:
>To be fair, I do handle garbage collection of deleting requests that have been taken care of, so I doubt even if this is used often that this object would get very large at all.

I'm not talking about garbage collection. I'm talking about polymorphism, which is something that deoptimizes code due to how V8 works

Zarel commented:
This is a pretty serious maintenance burden, incidentally, so if this is controversial among mods, maybe we shouldn't have this...

Morfent commented:
@Zarel do you know if Turbofan mitigates some of the issues with polymorphic/megamorphic code? I haven't looked into it much yet

Zarel commented:
It doesn't, as far as I know, but on the other hand, PS is already pretty polymorphic because we never really got around to caring about that...

Zarel commented:
I'm more worried about the general maintenance burden and the controversialness.

panpawn commented:
>I'm more worried about the general maintenance burden and the controversialness.

Based on the discussion in the policy thread, I really don't see this being much controversial at all...

AustinXII commented:
@panpawn there's multiple staff members who think this feature is not needed...

panpawn commented:
It seemed like there were significantly more people who thought PS could benefit from this than people who  were neutral towards it. If it really is that much of a controversy, I guess I'll just close this then...
